src/QsCompiler/Transformations/CallGraph/SimpleCallGraph.cs

```diff
@@ -0,0 +1,465 @@
+
+namespace Microsoft.Quantum.QsCompiler.DependencyAnalysis
+{
+    using Range = DataTypes.Range;
+    using TypeParameterResolutions = ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    /// <summary>
+    /// Edge type for Simple Call Graphs.
+    /// </summary>
+    public sealed class SimpleCallGraphEdge : CallGraphEdgeBase, IEquatable<SimpleCallGraphEdge>
+    {
+        /// <summary>
+        /// Contains the type parameter resolutions associated with this edge.
+        /// </summary>
+        public TypeParameterResolutions ParamResolutions { get; }
+
+        /// <summary>
+        /// Constructor for SimpleCallGraphEdge objects.
+        /// Strips position info from the given type parameter resolutions.
+        /// Throws an ArgumentNullException if any of the arguments are null.
+        /// </summary>
+        internal SimpleCallGraphEdge(TypeParameterResolutions paramResolutions, QsQualifiedName fromCallableName, QsQualifiedName toCallableName, Range referenceRange)
+            : base(fromCallableName, toCallableName, referenceRange)
+        {
+            if (paramResolutions is null)
+            {
+                throw new ArgumentNullException(nameof(paramResolutions));
+            }
+
+            // Remove position info from type parameter resolutions
+            this.ParamResolutions = paramResolutions.ToImmutableDictionary(
+                kvp => kvp.Key,
+                kvp => StripPositionInfo.Apply(kvp.Value));
+        }
+
+        /// <summary>
+        /// Determines if the object is the same as the given edge, ignoring the
+        /// ordering of key-value pairs in the type parameter dictionaries.
+        /// </summary>
+        public bool Equals(SimpleCallGraphEdge edge) =>
+            base.Equals(edge)
+            && (this.ParamResolutions == edge.ParamResolutions
+                || this.ParamResolutions
+                       .OrderBy(kvp => kvp.Key)
+                       .SequenceEqual(edge.ParamResolutions.OrderBy(kvp => kvp.Key)));
+    }
+
+    /// <summary>
+    /// Node type that represents Q# callables.
+    /// </summary>
+    public sealed class SimpleCallGraphNode : CallGraphNodeBase
+    {
+        /// <summary>
+        /// Constructor for SimpleCallGraphNode objects.
+        /// Throws an ArgumentNullException if the argument is null.
+        /// </summary>
+        public SimpleCallGraphNode(QsQualifiedName callableName)
+            : base(callableName)
+        {
+        }
+    }
+
+    /// <summary>
+    /// A kind of call graph whose nodes represent Q# callables.
+    /// </summary>
+    public sealed class SimpleCallGraph
+    {
+        // Static Elements
+
+        private static IEnumerable<IEnumerable<SimpleCallGraphEdge>> CartesianProduct(IEnumerable<IEnumerable<SimpleCallGraphEdge>> sequences)
+        {
+            IEnumerable<IEnumerable<SimpleCallGraphEdge>> result = new[] { Enumerable.Empty<SimpleCallGraphEdge>() };
```

Comment:
You could make this polymorphic since it doesn't need anything from the edge:

```suggestion
        private static IEnumerable<IEnumerable<T>> CartesianProduct<T>(IEnumerable<IEnumerable<T>> sequences)
        {
            IEnumerable<IEnumerable<T>> result = new[] { Enumerable.Empty<T>() };
```

Reply:
done